### ISSUE TITLE
Normal Kubernetes liveness probes prevent cluster bootstrap

### ISSUE THREAD

**Issue by patrick-premont** [akka/akka-management#259]
When starting a large number of Akka cluster nodes concurrently, that need to create a new cluster with the bootstrap mechanism, it may take more than a few minutes before discovery stabilizes, especially if most of those are created on a common Kubernetes node where there may be more resource contention.

If Kubernetes gives up any pod because of it failed too many liveness probes, it will restart it, likely preventing the bootstrap mechanism from ever converging: it could impact the discovered set, but mostly it will lead to contact point probing failures, and as long as there is any failure there, the bootstrap process will restart (As currently implemented and documented, bootstrap cannot succeed until  "Communication with all discovered Contact Points have been confirmed via successful HTTP request-response.") This will lead to more liveness probes expiring, preventing convergence.

Liveness probes do not appear to support activation only after the readiness probes have succeeded: https://github.com/kubernetes/kubernetes/issues/27114. So custom code would be needed to make them pass artificially until the cluster is formed.

Still, such a solution would not handle a scenario where one of the pods starts in a failed state where it is unable to respond to probe requests indefinitely. Could we relax the requirement that all contact points must respond?

If not, should we document that all automatic restarts must be prevented during the bootstrap process?

**Comment by ktoso** [comment]
Please clarify what ports this is probing

**Comment by ktoso** [comment]
Are we talking about the ones prepped by the rp lib etc?

**Comment by patrick-premont** [comment]
Yes, this is using the probes put in place by sbt-reactive-app.

**Comment by ktoso** [comment]
Yeah so not really about akka-management... but okey let's keep it here as I want to move those things as far out of sbt-reactive-app as possible :)

**Comment by patrick-premont** [comment]
The ability to delay liveness probes by some fixed amount of time slipped my mind (thanks @longmuir for reminding me of `initialDelaySeconds`). I was automatically using 60 seconds, as this was set by `reactive-cli`: https://github.com/lightbend/reactive-cli/blob/b6282d360d2704eb4f491cf118a4d509f23e2dc6/cli/shared/src/main/scala/com/lightbend/rp/reactivecli/runtime/kubernetes/package.scala#L37

Increasing this would be a good workaround, but perhaps we can find a more definitive solution that doesn't rely on such tuning.

**Comment by TimMoore** [comment]
This code lives in [`reactive-lib`](https://github.com/lightbend/reactive-lib) not in either `akka-management` or `sbt-reactive-app`. Specifically:

- base framework: https://github.com/lightbend/reactive-lib/tree/master/status/src/main/scala/com/lightbend/rp/status
- [`ClusterStatusCheck`](https://github.com/lightbend/reactive-lib/blob/master/akka-cluster-bootstrap/src/main/scala/com/lightbend/rp/akkaclusterbootstrap/ClusterStatusCheck.scala)

I think it could make sense to move it into Akka Management, however. Right now, it defines liveness and readiness identically.

After reading through the thread on kubernetes/kubernetes#27114, I think the best solution would be to redefine the liveness check.

The readiness check is used to determine whether a pod is ready to serve traffic, so it makes sense to return false when the cluster hasn't formed.

The liveness check is used to determine whether a pod is healthy or should be killed. This should be aware of the state of the node. A node that is still trying to bootstrap is in a healthy state (but is not yet ready). A node that has timed out and given up on bootstrapping, or one that has been downed, is unhealthy.

How does that sound, @ktoso and @patrick-premont? It might be more expedient to fix this in `reactive-lib` now, unless moving the code over to Akka Management is going to be done soon.

**Comment by patrick-premont** [comment]
I like that proposal @TimMoore. Want to propose this in a fresh issue in [reactive-lib](https://github.com/lightbend/reactive-lib) ?

I can open a separate issue for my remaining concern: that the requirement for hearing back from all discovered pods might be too strict.

Then we could be this one.
